src/main/java/org/craftercms/engine/util/servlet/ConfigAwareProxyServlet.java

```diff
@@ -84,4 +111,155 @@ protected void copyResponseHeaders(HttpResponse proxyResponse, HttpServletReques
         }
     }
 
+    @Override
+    protected void service(HttpServletRequest servletRequest, HttpServletResponse servletResponse)
+            throws IOException, ServletException {
+        var websocketKey = servletRequest.getHeader(WEBSOCKET_IDENTIFIER_HEADER);
+        if (websocketKey != null) {
+            websocketHandlerService(servletRequest, servletResponse);
+        } else {
+            super.service(servletRequest, servletResponse);
+        }
+    }
+
+    /**
+     * Handle the initial handshake request for websocket connection and establish the websocket.
+     * A WebSocket interaction begins with an HTTP request that uses the HTTP Upgrade header to upgrade to the WebSocket protocol.
+     * This method first forward the handshake request to the websocket server and read the response.
+     * If the response is 101, use {@link WsUpgradeHandler} to establish the websocket proxy among client <==> engine <==> websocket server.
+     * @param servletRequest servlet request from client
+     * @param servletResponse servlet response to client
+     */
+    private void websocketHandlerService(HttpServletRequest servletRequest, HttpServletResponse servletResponse) throws IOException, ServletException {
+        //initialize request attributes from caches if unset by a subclass by this point
+        if (servletRequest.getAttribute(ATTR_TARGET_URI) == null) {
+            servletRequest.setAttribute(ATTR_TARGET_URI, targetUri);
+        }
+        if (servletRequest.getAttribute(ATTR_TARGET_HOST) == null) {
+            servletRequest.setAttribute(ATTR_TARGET_HOST, targetHost);
+        }
+        String proxyRequestUri = rewriteUrlFromRequest(servletRequest);
+        URL url = new URL(proxyRequestUri);
+
+        // Create a new socket connection to the websocket server
+        Socket socketProxyClient = new Socket(url.getHost(), url.getPort());
+        boolean closeSocket = false;
+        try {
+            var socketIn = socketProxyClient.getInputStream();
+            var socketOut = socketProxyClient.getOutputStream();
+
+            // Send the handshake request from engine to the websocket server
+            String handshakeRequest = getProxyHandshakeRequest(servletRequest, url);
+            socketOut.write(handshakeRequest.getBytes(StandardCharsets.UTF_8));
```

Comment:
This specific method is only for the *HTTP handshake* request, which there is no body. A sample of such request:

```
curl 'ws://localhost:8080/_next/webpack-hmr' \
  -H 'Pragma: no-cache' \
  -H 'Origin: http://localhost:8080' \
  -H 'Accept-Language: en-US,en;q=0.9,vi;q=0.8' \
  -H 'Sec-WebSocket-Key: YDqLOWPZ8OSKy148rNQIEQ==' \
  -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36' \
  -H 'Upgrade: websocket' \
  -H 'Cache-Control: no-cache' \
  -H 'Cookie: xxxxxxxxxxxxxx \
  -H 'Connection: Upgrade' \
  -H 'Sec-WebSocket-Version: 13' \
  -H 'Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits' \
  --compressed
```

I believe we are doing the same as the current proxy library is doing: Reading all headers into memory and then forward to the server: https://github.com/mitre/HTTP-Proxy-Servlet/blob/master/src/main/java/org/mitre/dsmiley/httpproxy/ProxyServlet.java#L490.

The socket should not be closed here since we will handle the socket connection with `WsUpgradeHandler` after this initial handshake HTTP request responds with status `101`. If the status is not `101`, we close the socket immediately: https://github.com/craftercms/engine/pull/682/files#diff-44b6059479d4bd49c012bbcd46e18e48b5c2281d8671873ee6cf2ce6e524021aR170

Reply:
I did some basic research, and it seems that:
* There is no limit on the size of transmitted HTTP headers in the HTTP spec
* Apache Tomcat and other HTTP servers ahead of us will limit the HTTP headers size

The second point makes me feel better and we can proceed with this PR.

Reference: https://stackoverflow.com/questions/686217/maximum-on-http-header-values